internal/blockchain/utxoviewpoint.go

```diff
@@ -216,26 +216,41 @@ func (view *UtxoViewpoint) PriorityInput(prevOut *wire.OutPoint) (int64, int64,
 	return 0, 0, false
 }
 
-// connectTransaction updates the view by adding all new utxos created by the
-// passed transaction and marking all utxos that the transaction spends as
-// spent.  In addition, when the 'stxos' argument is not nil, it will be updated
-// to append an entry for each spent txout.  An error will be returned if the
-// view does not contain the required utxos.
-func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
-	blockIndex uint32, stxos *[]spentTxOut, isTreasuryEnabled bool) error {
-
-	// Treasurybase transactions don't have any inputs to spend.
-	//
-	// NOTE: This check MUST come before the coinbase check because a
-	// treasurybase is identified as a coinbase as of the time this comment is
-	// being written.
+// utxoEntryToSpentTxOut converts the provided utxo entry to a spent txout for
+// use in tracking all spent txouts for the spend journal.
+func utxoEntryToSpentTxOut(entry *UtxoEntry) spentTxOut {
+	// Populate the stxo details using the utxo entry.
+	return spentTxOut{
+		amount:        entry.Amount(),
+		pkScript:      entry.PkScript(),
+		ticketMinOuts: entry.ticketMinOuts,
+		blockHeight:   uint32(entry.BlockHeight()),
+		blockIndex:    entry.BlockIndex(),
+		scriptVersion: entry.ScriptVersion(),
+		packedFlags: encodeFlags(entry.IsCoinBase(), entry.HasExpiry(),
+			entry.TransactionType()),
+	}
+}
+
+// connectStakeTransaction updates the view by adding all new utxos created by
+// the passed transaction from the stake transaction tree and marking all utxos
+// that the transaction spends as spent.  In addition, when the 'stxos' argument
+// is not nil, it will be updated to append an entry for each spent txout.  An
+// error will be returned if the view does not contain the required utxos.
+func (view *UtxoViewpoint) connectStakeTransaction(tx *dcrutil.Tx,
+	blockHeight int64, blockIndex uint32, stxos *[]spentTxOut,
+	isTreasuryEnabled bool) error {
+
+	// Treasurybase transactions don't have any inputs to spend or outputs to
+	// add.
 	msgTx := tx.MsgTx()
 	if isTreasuryEnabled && standalone.IsTreasuryBase(msgTx) {
```

Comment:
I'm curious about the use of `IsTreasuryBase` instead of `blockIndex==0`.  Seems like more work than needed.

Reply:
Discussed in the other related comment.

Reply:
I updated this as follows:

```Go
	//
	// NOTE: This check relies the code already having validated that the first
	// transaction, and only the first transaction, in the stake tree is the
	// required treasurybase when the treasury agenda is active.
	msgTx := tx.MsgTx()
	if isTreasuryEnabled && blockIndex == 0 {
```

It's a lot safer to rely on this assumption here than the other case, because this case is required to be much deeper in the consensus checks which means it is exceedingly unlikely the code could be changed in such a way as to violate it.